Skip intro and ending per feed

tonytamsf opened the pull request:
fix #1552

I'm ready for a code review.

On a per feed basis
1) Allow the user to set skip intro and skip ending dialog in preferences
2) When the usrs plays an episode, show them a toast message that we are skipping ahead and marking ending early
3) When we are ending early, give them a toast message as well

Feedback needed:
1) Are there too many toast messages?
2) I added 2 more fields into the database.  What do I need to do to make sure there isn't any issues for people to import and export databases when they upgrade?  Take a look at the debug log https://gist.github.com/tonytamsf/e704090f5031e58da240ea976da8d1f5
2.1) looks like this upgrade [code worked](https://github.com/AntennaPod/AntennaPod/pull/3975/commits/b93bce74012b31980fe3a1e19a49df82901e1978)

https://github.com/AntennaPod/AntennaPod/pull/3975
![ezgif com-video-to-gif(1)](https://user-images.githubusercontent.com/149837/77842715-786ab880-714a-11ea-94ca-4ab0fb3ed133.gif)

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/LocalPSMP.java:
```diff
@@ -317,12 +320,53 @@ private void resumeSync() {
+                        int duration = getDuration();
+                        if (skipIntro * 1000 < duration) {
+                            seekToSync(skipIntro * 1000);
+                            skipIntroMesg = context.getString(R.string.pref_feed_skip_intro) + " " +
```
Building Strings like this can break in other languages. I think an additional string `Skipped %1$d seconds of intro` would be better.

tonytamsf replied:
done

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/LocalPSMP.java:
```diff
@@ -317,12 +320,53 @@ private void resumeSync() {
+                        }
+                    }
+                    if (skipMesg != "") {
+                        skipMesg = skipIntroMesg + "\n" + skipMesg;
```
Why do you display both numbers? I think we should not display more information than necessary.

tonytamsf replied:
removed dup

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/storage/PodDBAdapter.java:
```diff
@@ -1381,6 +1389,12 @@ public void onCreate(final SQLiteDatabase db) {
         public void onUpgrade(final SQLiteDatabase db, final int oldVersion, final int newVersion) {
             Log.w("DBAdapter", "Upgrading from version " + oldVersion + " to " + newVersion + ".");
             DBUpgrader.upgrade(db, oldVersion, newVersion);
+            if (oldVersion == 1090000) {
```
Please move this to `DBUpgrader`. Also, use `oldVersion < 1090001`. Users might skip AntennaPod versions. When checking for equality, this skips the migration when coming from, say `1080100` instead of exactly `1090000`.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/storage/PodDBAdapter.java:
```diff
@@ -48,7 +48,7 @@
     private static final String TAG = "PodDBAdapter";
     public static final String DATABASE_NAME = "Antennapod.db";
-    public static final int VERSION = 1090000;
+    public static final int VERSION = 1091000;
```
Please only increment by 1.

tonytamsf replied:
done `1090000` + 1 -> 1090001

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/LocalPSMP.java:
```diff
@@ -317,12 +320,53 @@ private void resumeSync() {
+                    if (skipIntro > 0) {
+                        int duration = getDuration();
+                        if (skipIntro * 1000 < duration) {
+                            seekToSync(skipIntro * 1000);
```
Not sure if `LocalPSMP` is the right place to do this. What about Chromecast (`RemotePSMP`) or other media players that might get added in the future? I think this fits better into `PlaybackService` (but I am not sure where exactly).

tonytamsf replied:
fixed in Playbackservice

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/AudioPlayerFragment.java:
```diff
@@ -386,6 +390,24 @@ public void updatePosition(PlaybackPositionEvent event) {
+                Toast toast = Toast.makeText(getActivity(), skipMesg,
+                        Toast.LENGTH_LONG);
+                toast.show();
+                IntentUtils.sendLocalBroadcast(getActivity(), PlaybackService.ACTION_SKIP_CURRENT_EPISODE);
```
The UI should not be responsible for doing playback actions. If the user closes the UI, episodes should still get skipped. When adding the skipping code like this and swiping out the player during playback, this is no longer the case.

Additionally, this introduces code duplication. This should be done centrally, somewhere in the PlaybackService.

tonytamsf replied:
done.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/FeedPreferenceSkipDialog.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public abstract class FeedPreferenceSkipDialog extends AlertDialog.Builder {
+
+    public FeedPreferenceSkipDialog(Context context, int titleRes, int skipIntroInitialValue,
```
Is there a case where the title might be different from `R.string.pref_feed_skip`? I do not think this needs to be a parameter.

tonytamsf replied:
done

ByteHamster commented on app/src/main/res/layout/feed_pref_skip_dialog.xml:
```diff
@@ -0,0 +1,74 @@
+        android:layout_marginRight="24dp"
+        android:text="@string/pref_feed_skip_intro" />
+
+    <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
This file could need some code cleanup. There are some duplicated namespace declarations, indentation is not consistent and the number of empty lines between Views is not consistent. `Ctrl+Alt+Shift+L` in Android Studio could help.

tonytamsf replied:
done

tonytamsf commented:
@ByteHamster please do a review again for all my changes.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -505,6 +507,36 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+
+        int skipIntro = 0;
+        if (! (playable instanceof FeedMedia)) { return; }
+        if (playable.getLastPlayedTime() > 0) { return; }
```
There is a checkstyle file in config/checkstyle/ that I want to enforce for new code. Because the existing code is quite inconsistent, CI only checks a small fraction of things. In the past months, I did the checks manually and commented on GitHub. In #3978, I enabled checking the diff on CI.

I think checkstyle will complain about this line. Have not checked the rest of the code. Just that you already know before I merge #3978 and it suddenly complains.

tonytamsf replied:
not sure what it should be instead?
```
if (playable.getLastPlayedTime() > 0) { return; }
```

tonytamsf replied:
@ByteHamster could you check the checkstyle, it looks like the error is not related to my changes or the reporting is not super clear

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/activity/MediaplayerActivity.java:
```diff
@@ -477,6 +480,24 @@ void onPositionObserverUpdate() {
         }
         updateProgressbarPosition(currentPosition, duration);
+        Playable media = controller.getMedia();
+        if (media instanceof FeedMedia)  {
```
This is no longer needed, I think.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1071,32 @@ private void sendNotificationBroadcast(int type, int code) {
+        if (skipEnd > 0 && remainingTime < skipEnd * 1000) {
+            Log.d(TAG, "skipIfMetEnding: Skipping the remaining " + remainingTime / 1000 + " " + skipEnd );
+            Context context = getApplicationContext();
+            String skipMesg = String.format(context.getString(R.string.pref_feed_skip_ending_toast),
```
I think you can directly add the format parameters to getString. Then Android Studio can do better analysis.

tonytamsf replied:
fixed

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1071,32 @@ private void sendNotificationBroadcast(int type, int code) {
+                    Toast.LENGTH_LONG);
+            toast.show();
+
+            mediaPlayer.skip();
```
Just wondering: Does this honor the "Follow queue" setting?

tonytamsf replied:
I tested it and this behaves the same as hitting the 'Skip' button
1) The location is saved
2) The video is skipped and the next one is started.

I don't see 'Follow queue'  option

ByteHamster replied:
> The video is skipped and the next one is started.

What if you uncheck the option `Follow queue` in Settings » Playback? It should then not start the next file.

tonytamsf replied:
fixed now and tested.

tonytamsf commented on core/src/main/java/de/danoeh/antennapod/core/storage/PodDBAdapter.java:
```diff
@@ -145,7 +147,9 @@
-            + KEY_FEED_VOLUME_ADAPTION + " INTEGER DEFAULT 0)";
+            + KEY_FEED_VOLUME_ADAPTION + " INTEGER DEFAULT 0,"
+            + KEY_FEED_SKIP_INTRO + " INTEGER DEFAULT 0,"
+            + KEY_FEED_SKIP_ENDING+ " INTEGER DEFAULT 0)";
```
done

tonytamsf commented on core/src/main/java/de/danoeh/antennapod/core/feed/FeedPreferences.java:
```diff
@@ -15,6 +15,8 @@
 
     public static final float SPEED_USE_GLOBAL = -1;
+    public static final int FEED_SKIP_INTRO_GLOBAL = 0;
+    public static final int FEED_SKIP_ENDING_GLOBAL = 0;
```
done

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/dialog/FeedPreferenceSkipDialog.java:
```diff
@@ -0,0 +1,35 @@
+package de.danoeh.antennapod.dialog;
+
+import android.app.AlertDialog;
```
Please use the `androidx` alert dialog. It has the same interface but it might have better compatibility with old Android versions.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/feed/FeedPreferences.java:
```diff
@@ -34,12 +34,17 @@
     private FeedPreferences(long feedID, boolean autoDownload, boolean keepUpdated, AutoDeleteAction auto_delete_action, VolumeAdaptionSetting volumeAdaptionSetting, String username, String password, @NonNull FeedFilter filter, float feedPlaybackSpeed) {
+        this(feedID, autoDownload, true, auto_delete_action, volumeAdaptionSetting, username, password, new FeedFilter(), feedPlaybackSpeed, 0, 0);
+    }
+    private FeedPreferences(long feedID, boolean autoDownload, boolean keepUpdated, AutoDeleteAction auto_delete_action, VolumeAdaptionSetting volumeAdaptionSetting, String username, String password, @NonNull FeedFilter filter, float feedPlaybackSpeed, int feedSkipIntro, int feedSkipEnding) {
```
Please add an empty line between the methods. Not sure why checkstyle did not complain about that. The checkstyle config file needs some more changes, I think.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -505,6 +507,40 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+            return;
+        }
+
+        if (playable.getLastPlayedTime() > 0) {
```
Doesn't this mean that the intro is not skipped if you play it for the second time?

tonytamsf replied:
I have fixed.
1) If you play, autoskip and rewind, it works as you expected.
2) If you play, autoskip, let it end. and come back, it will work as you expect.  The 2nd time, it will skip again

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -505,6 +507,40 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+                        context.getString(R.string.time_seconds));
+            }
+        }
+        if (skipIntroMesg != "") {
```
I would move this up into the two `if`s above. Then you can skip the check (`skipIntroMesg != ""` if and only if the code above is executed).

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1075,34 @@ private void sendNotificationBroadcast(int type, int code) {
         sendBroadcast(intent);
     }
 
+    private void skipIfMetEnding() {
```
`positionSaverTick` is only executed every 5 seconds. I think you can use EventBus:

```
@Subscribe
public void skipIfMetEnding(PlaybackPositionEvent event) {
```

tonytamsf replied:
instead of a new Subscriber, I'm piggybacking on the onPositionObserver()

ByteHamster commented on core/src/main/res/values/strings.xml:
```diff
@@ -449,6 +449,12 @@
+    <string name="pref_feed_skip_sum">Skip introductions and ending credits.</string>
+    <string name="pref_feed_skip_ending">Skip last</string>
+    <string name="pref_feed_skip_intro">Skip first</string>
+    <string name="pref_feed_skip_ending_toast">Skip last %d %s</string>
```
I would use `Skipped` instead of `Skip` in the Toasts because they show something that happened, not an action that can be done.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -496,6 +497,7 @@ public int onStartCommand(Intent intent, int flags, int startId) {
                     return Service.START_NOT_STICKY;
                 }
                 mediaPlayer.playMediaObject(playable, stream, startWhenPrepared, prepareImmediately);
+                skipIntro(playable);
```
I don't think that this is enough. If AntennaPod automatically plays the next item in the queue, this code is not executed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1068,34 @@ private void sendNotificationBroadcast(int type, int code) {
+            Context context = getApplicationContext();
+            String skipMesg = context.getString(R.string.pref_feed_skip_ending_toast,
+                    skipEnd,
+                    context.getString(R.string.time_seconds));
```
I think it would be easier for the translators if you just added the text "Seconds" to `strings.xml`. I do not think that we want to skip other units than seconds. Additionally, this gives a weird mix of languages if "Seconds" is translated but "Skipped first" is not translated.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -505,6 +506,34 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+            return;
+        }
+
+        if (playable instanceof FeedMedia) {
```
This is always true. You check for FeedMedia in the line above.

tonytamsf commented:
> I discovered a few edge cases:

Fantastic testing

>     * If you manually press play on an episode where the position already is after `skipEnd`, it should not skip. This might happen if you decide to listen to the rest of the episode after skipping.

Thinking about how this should work.  Because of playback speed preference, I cannot just check for the exact second when the playback happens and only skip at that particular second.  I could look at if you scrub pass the skip, calculate roughly where you when pass the skipEnd and don't skip if your intention is to play the ending.  *and* I might have to visually denote where the skip part is in the timeline?

>     * If you enter an incredibly high number (or have a keyboard that still allows entering text), AntennaPod crashes with a `java.lang.NumberFormatException`.

Fixed
 
>     * If the skipEnd time is longer than the episode, the episode is not played at all

Fixed

>     * Episodes should be marked as played when skipping the last seconds. This depends on the "Smart mark as played" feature. I think when skipping the ending automatically, the episode should always be marked as played, removed from queue, etc (probably needs changes in `onPostPlayback`)

It does get marked as played and removed from queue if you have the "Keep Skipped Episodes" *unchecked*.  The .skip() I'm using respects the same thing as if you hit the 'skip' button.  What do you think?

>     * When having the "Continuous playback" setting enabled and playback goes to the next episode in the queue, the next episode's beginning does not seem to be skipped

I have tested with episodes in the queue and the next episode is skipped correct.  Are you sure?

ByteHamster commented:
> It does get marked as played and removed from queue if you have the "Keep Skipped Episodes" unchecked. The .skip() I'm using respects the same thing as if you hit the 'skip' button. What do you think?

I think this setting should apply to manual skips only. If you select to automatically skip the last 1 minute but "Keep Skipped Episodes" is on (default), it would not mark as played. I think "Skip last x seconds", to the user, does not feel like skipping. It feels like trimming the audio file, so automatically skipped episodes should still be marked as played (synced to Gpodder etc), regardless of the setting. One way to fix this would be to have `boolean smartMarkAsPlayed = media.hasAlmostEnded();` in `onPostPlayback` consider the skip interval.

> I have tested with episodes in the queue and the next episode is skipped correct. Are you sure?

Okay, you are right. I think I have experienced something related to #3892 where the stored position was not updated correctly.

> Thinking about how this should work.

Maybe some kind of flag that disables automatic skipping for that one episode? If the episode is started after the skip interval, do not do automatic skipping. Maybe even if a user seeks behind the skip interval manually, the flag could be set. I am not sure if this requires massive code changes, though.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedSettingsFragment.java:
```diff
@@ -159,6 +161,37 @@ public void onDestroy() {
+                            skipIntro = Integer.parseInt(skipIntroStr);
+                        } catch (NumberFormatException e) {
+                            skipIntro = 0;
+                        }
```
I think this could be done inside `FeedPreferenceSkipDialog` to keep the `FeedSettingsFragment` a bit shorter

tonytamsf replied:
done

tonytamsf commented:
> I think this setting should apply to manual skips only. If you select to automatically skip the last 1 minute but "Keep Skipped Episodes" is on (default), it would not mark as played. I think "Skip last x seconds", to the user, does not feel like skipping. It feels like trimming the audio file, so automatically skipped episodes should still be marked as played (synced to Gpodder etc), regardless of the setting. One way to fix this would be to have `boolean smartMarkAsPlayed = media.hasAlmostEnded();` in `onPostPlayback` consider the skip interval.

I've made the changes.  Great suggestion.  I've been using this feature daily now.  I like how the skipping is intuitive, (what I would expect)

tonytamsf commented:
> I think the only bug I have left is when you have the episode in the queue, you already set skipEnd preference to X, and when you set to Y, the data is modified but I will still use the older value in X. Looks like I might need to use Event and Subscribe?

I've fixed in `d41ede9` - Learning something new about [EventBus](https://greenrobot.org/eventbus/documentation/faq/). Thank you!

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1070,37 @@ private void sendNotificationBroadcast(int type, int code) {
         sendBroadcast(intent);
     }
 
+    private void skipEnding() {
```
What about `skipEndingIfNecessary`?

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1070,37 @@ private void sendNotificationBroadcast(int type, int code) {
+            playable.setAutoSkippedEnding(true);
+            mediaPlayer.skip();
+        } else {
+           Log.d(TAG, "skipEnding: Not at ending yet " + remainingTime + " diff " + (remainingTime - (skipEnd * 1000)) + " skipEnd " + skipEnd * 1000 + " speed " + getCurrentPlaybackSpeed() * 1000);
```
Please remove this. It is executed every second, so this floods the log output

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1070,37 @@ private void sendNotificationBroadcast(int type, int code) {
+        FeedPreferences preferences = feedMedia.getItem().getFeed().getPreferences();
+        int skipEnd = preferences.getFeedSkipEnding();
+        if (skipEnd > 0 &&
+            skipEnd < playable.getDuration() &&
```
This PR does not include the more strict diff-checkstyle CI rule. The `&&` characters should be in the next line. Also, indentation should be 8 spaces when breaking the same statement.

You can select the file `config/checkstyle/checkstyle-new-code.xml` in IntelliJ/Android Studio (with checkstyle version 8.18 and "Treat errors as warnings") to get inline suggestions. Alternatively, you can rebase the commit with a more recent `develop` branch to have it executed on CI.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1037,6 +1070,37 @@ private void sendNotificationBroadcast(int type, int code) {
+            String skipMesg = context.getString(R.string.pref_feed_skip_ending_toast,
+                    skipEnd);
+            Toast toast = Toast.makeText(context, skipMesg,
+                    Toast.LENGTH_LONG);
```
Could you please make above statements a single line each? There is a lot of horizontal screen space left :)

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -997,6 +1026,10 @@ private void onPostPlayback(final Playable playable, boolean ended, boolean skip
             media.onPlaybackPause(getApplicationContext());
         }
 
+        if (playable.hasAutoSkippedEnding()) {
```
Hmm this adds quite a lot of code for a single use-case. What about checking the position here instead of adding the Playable methods?

tonytamsf commented:
> Have not tested the code but I think this one is still open. I would change PlaybackService to have a boolean flag somewhere that tells if skipping the ending is enabled for the current file. Every time a new file is played, the boolean is changed according to the initial playback position.

Do I need to keep state of this for when the app is closed and started again?  Right now, I'm assuming if you scrub pass the (playback speed * 1 sec) when you want to continue.  Could you test it out and see if it behaves like you expect it?

tonytamsf commented:
> Sorry for being so picky... I have another big round of change requests 🙈

No worries, I'm learning through your code reviews.  I hope to make this feature into 2.0

tonytamsf commented:
@ByteHamster are you still testing this feature?

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/feed/FeedPreferences.java:
```diff
@@ -49,6 +55,8 @@ private FeedPreferences(long feedID, boolean autoDownload, boolean keepUpdated,
         this.password = password;
         this.filter = filter;
         this.feedPlaybackSpeed = feedPlaybackSpeed;
+        this.feedSkipIntro= feedSkipIntro;
```
Missing space :)

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/feed/FeedPreferences.java:
```diff
@@ -204,4 +227,12 @@ public float getFeedPlaybackSpeed() {
         feedPlaybackSpeed = playbackSpeed;
     }
+
+    public void setFeedSkipIntro(int skipIntro) { feedSkipIntro = skipIntro; }
```
Only one statement per line, please

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -210,6 +212,8 @@
     private Disposable positionEventTimer;
     private PlaybackServiceNotificationBuilder notificationBuilder;
 
+    FeedMedia autoSkippedFeedMedia;
```
Can this be private? I think it would be enough to only save the ID instead of the whole object.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedSettingsFragment.java:
```diff
@@ -159,6 +162,26 @@ public void onDestroy() {
         }
 
+        private void setupFeedAutoSkipPreference() {
+            findPreference("skipping").setOnPreferenceClickListener(preference -> {
```
Sorry, I just noticed this line (otherwise, I would have written that earlier). Please use a constant here, just like the other preferences in this class.

tonytamsf replied:
No worries will change

tonytamsf commented:
I will try to write a test, I wanted to learn how to do that anyways.

ByteHamster commented:
> I will try to write a test, I wanted to learn how to do that anyways.

Do you prefer to merge as-is (tests in another PR) or do you want to add the tests here?

tonytamsf commented:
@ByteHamster please merge as is, it is taking me longer to get the test in there.

rulnick commented:
Thank you for all of your work on this.

Erwan62 commented:
Où en est cette fonctionnalité ?
